app/src/main/java/org/mozilla/focus/browser/BrowserNavigationOverlay.kt

```diff
@@ -151,10 +161,28 @@ class BrowserNavigationOverlay @JvmOverloads constructor(
             else -> R.id.navButtonForward
         }
 
-        navUrlInput.setText(navigationStateProvider?.getCurrentUrl())
+        maybeUpdateOverlayURLForCurrentState()
```

Comment:
> When the overlay is closed, I assume the urlbar's onFocusChangedListener is also called? 

Yes. In order for the user to interact with the view, they need focus. But I suppose this could be fragile if we end up re-using this View and focus isn't re-used for some reason. `onFocus == true` would be safer to reinit, but I'm concerned that could get called multiple times while the view is focused... (but maybe that wouldn't happen without `onFocusChanged(false)` first.

---

Bonus high-level concept: taking a step back, the other day we were talking about how one component of React is a "render function": you take all of the app's state and you re-render the UI from that state each time any part of the state changes.

This function acts as "render()": it takes all of the state it knows about (as member variables) and updates the UI accordingly.

> this could be called from the BrowserFragment (when the loading state changes) or the Overlay itself

In theory, it can be called from anywhere and it'd be accurate (ignoring concurrency problems and how it could interrupt the usere)!